pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -735,3 +738,71 @@ func getHashPolicy(push *model.PushContext, dst *networking.DestinationWeight) *
 
 	return nil
 }
+
+type envoyRouteType int
+
+const (
+	envoyPath envoyRouteType = iota
+	envoyPrefix
+	envoyRegex
+	envoyCatchAll
+)
+
+func getEnvoyRouteTypeAndVal(r *route.Route) (envoyRouteType, string) {
+	var iType envoyRouteType
+	var iVal string
+
+	switch iR := r.Match.PathSpecifier.(type) {
+	case *route.RouteMatch_Path:
+		iVal = iR.Path
+		iType = envoyPath
+	case *route.RouteMatch_Prefix:
+		iVal = iR.Prefix
+		iType = envoyPrefix
+	case *route.RouteMatch_Regex:
+		iVal = iR.Regex
+		iType = envoyRegex
+	}
+
+	// A route is catch all if and only if it has no header/query param match
+	// and has a prefix / or regex *.
+	if (iVal == "/" && iType == envoyPrefix) || (iVal == "*" && iType == envoyRegex) {
```

Comment:
If we want do the absolute minimum, to keep all options open for how things are reordered in the future, I would even reduce this check to only the prefix / case. It's the legitimate VirtualService-without-a-match-clause catch-all case.

Reply:
Not necessarily. User could also create a match clause with just prefix /

Reply:
good point. bottom line is moving these rules could cause breakage, but I think it's a pretty unlikely corner case.